modular_skyrat/modules/modular_implants/code/soulcatcher/soulcatcher_component.dm

```diff
@@ -0,0 +1,273 @@
+	soulcatcher_rooms += created_room
+
+	created_room.master_soulcatcher = WEAKREF(src)
+
+///Recieves a message from a soulcatcher room.
+/datum/component/soulcatcher/proc/recieve_message(message_to_recieve)
+	if(!message_to_recieve)
+		return FALSE
+
+	if(istype(parent, /obj/item))
+		var/obj/item/parent_item = parent
+		var/mob/living/holder = parent_item.loc
+		if(!holder)
+			return FALSE
+
+		to_chat(holder, message_to_recieve)
+		return TRUE
+
+	if(istype(parent, /obj/item/organ/internal/cyberimp/brain/nif))
+		var/obj/item/organ/internal/cyberimp/brain/nif/target_nif = parent
+		to_chat(target_nif.linked_mob, message_to_recieve)
+		return TRUE
+
+	return FALSE
+
+/datum/soulcatcher_room
+	/// What is the name of the room?
+	var/name = "Default Room"
+	/// What is the description of the room?
+	var/room_description = "An orange platform suspended in space orbited by reflective cubes of various sizes. There really isn't much here at the moment."
+	/// What souls are currently inside of the room?
+	var/list/current_souls = list()
+	/// Weakref for the master soulcatcher datum
+	var/datum/weakref/master_soulcatcher
+	/// What is the name of the person sending the messages?
+	var/outside_voice = "Host"
+	/// Can the room be joined at all?
+	var/joinable = TRUE
+
+/// Attemps to add a ghost to the soulcatcher room.
+/datum/soulcatcher_room/proc/add_soul_from_ghost(mob/dead/observer/ghost)
+	if(!ghost || !ghost.ckey)
+		return FALSE
+
+	if(!ghost.mind)
+		ghost.mind = new /datum/mind(ghost.key)
+		ghost.mind.name = ghost.name
+		ghost.mind.active = TRUE
+
+	if(!add_soul(ghost.mind))
+		return FALSE
+
+	return TRUE
+
+/// Converts a mind into a soul and adds the resulting soul to the room.
+/datum/soulcatcher_room/proc/add_soul(datum/mind/mind_to_add)
+	if(!mind_to_add)
+		return FALSE
+
+	var/datum/component/soulcatcher/parent_soulcatcher = master_soulcatcher.resolve()
+	var/datum/parent_object = parent_soulcatcher.parent
+	if(!parent_object)
+		return FALSE
+
+	var/mob/living/soulcatcher_soul/new_soul = new(parent_object)
+	new_soul.name = mind_to_add.name
+
+	if(mind_to_add.current)
+		var/datum/component/previous_body/body_component = mind_to_add.current.AddComponent(/datum/component/previous_body)
+		body_component.soulcatcher_soul = WEAKREF(new_soul)
+
+		new_soul.round_participant = TRUE
+		new_soul.body_scan_needed = TRUE
+
+		new_soul.previous_body = WEAKREF(mind_to_add.current)
+		new_soul.name = pick(GLOB.last_names) //Until the body is discovered, the soul is a new person.
+		new_soul.soul_desc = "[new_soul] lacks a discernible form."
+
+	mind_to_add.transfer_to(new_soul, TRUE)
+	current_souls += new_soul
+	new_soul.current_room = WEAKREF(src)
+
+	var/datum/preferences/preferences = new_soul.client?.prefs
+	if(preferences)
+		new_soul.ooc_notes = preferences.read_preference(/datum/preference/text/ooc_notes)
+		if(!new_soul.body_scan_needed)
+			new_soul.soul_desc = preferences.read_preference(/datum/preference/text/flavor_text)
```

Comment:
Isn't this a potential issue when it comes to getting info about someone that's died in the round?

Reply:
no